BUG: TimedCache.Invalidate() crashes with "close of closed channel" panic

# Description

`TimedCache.Invalidate()` in `core/pkg/policyhandler/cache.go` calls `close(c.stopChan)` without checking whether the channel is already closed. This causes a panic with the message "close of closed channel", crashing the process.

The panic happens in three scenarios:

1. `Invalidate()` is called twice in a row
2. `Stop()` is called before `Invalidate()`
3. `Invalidate()` and `Stop()` are called at the same time

Both `Stop()` and `Invalidate()` close `c.stopChan`, and neither checks if the channel is already closed. Calling `Stop()` also does not wait for the `invalidateTask()` goroutine to finish, so that goroutine is left hanging.

# Environment
OS: macOS 
Version: `dev`

# Steps To Reproduce

1. Create a new timed cache with a TTL, for example: `cache := NewTimedCache[string](1 * time.Hour)`
2. Set a value in the cache: `cache.Set("test")`
3. Call `cache.Stop()`
4. Call `cache.Invalidate()`
5. The program crashes with `panic: close of closed channel` at line 106 in `core/pkg/policyhandler/cache.go`

# Expected behavior

Calling `Stop()` and `Invalidate()` in any order should not crash the program. If `Invalidate()` is called after `Stop()`, it should simply do nothing rather than panic. The `invalidateTask()` goroutine should also exit cleanly when `Stop()` is called.

# Actual Behavior

The program crashes with this panic:

```
panic: close of closed channel
```

--- a/core/pkg/policyhandler/cache.go
+++ b/core/pkg/policyhandler/cache.go
@@ -2,21 +2,24 @@ package policyhandler
 
 import (
 	"sync"
+	"sync/atomic"
 	"time"
 )
 
-// TimedCache provides functionality for managing a timed cache.
-// The timed cache holds a value for a specified time duration (TTL).
-// After the TTL has passed, the value is invalidated.
-//
-// The cache is thread safe.
+const (
+	cacheStateActive uint32 = iota
+	cacheStateStopping
+)
+
 type TimedCache[T any] struct {
 	value      T
 	isSet      bool
 	ttl        time.Duration
 	expiration time.Time
 	mutex      sync.RWMutex
-	stopChan   chan struct{} // to stop the invalidateTask goroutine
+	stopChan   chan struct{}
+	stopWg     sync.WaitGroup
+	stopped    uint32
 }
 
 func NewTimedCache[T any](ttl time.Duration) *TimedCache[T] {
@@ -26,8 +29,8 @@ func NewTimedCache[T any](ttl time.Duration) *TimedCache[T] {
 		stopChan: make(chan struct{}),
 	}
 
-	// start the invalidate task only when the ttl is greater than 0 (cache is enabled)
 	if ttl > 0 {
+		cache.stopWg.Add(1)
 		go cache.invalidateTask()
 	}
 
@@ -46,9 +49,8 @@ func (c *TimedCache[T]) Set(value T) {
 	c.value = value
 	c.expiration = time.Now().Add(c.ttl)
 
-	// Signal invalidation to Get() if cache is already expired
 	if time.Now().After(c.expiration) {
-		c.Invalidate()
+		c.invalidateLocked()
 	}
 }
 
@@ -56,21 +58,19 @@ func (c *TimedCache[T]) Get() (T, bool) {
 	c.mutex.RLock()
 	defer c.mutex.RUnlock()
 
-	// If the invalidateTask() goroutine is currently invalidating the cache,
-	// the Get() method may return the stale cached value before the invalidation is complete.
-	// To avoid the stale cached value, we're requiring the Get() method to wait for the invalidation signal before returning the value.
-	select {
-	case <-c.stopChan:
+	if atomic.LoadUint32(&c.stopped) != cacheStateActive {
 		return c.value, false
-	default:
-		if !c.isSet || time.Now().After(c.expiration) {
-			return c.value, false
-		}
-		return c.value, true
 	}
+
+	if !c.isSet || time.Now().After(c.expiration) {
+		return c.value, false
+	}
+	return c.value, true
 }
 
 func (c *TimedCache[T]) invalidateTask() {
+	defer c.stopWg.Done()
+
 	ticker := time.NewTicker(c.ttl)
 	defer ticker.Stop()
 
@@ -81,28 +81,38 @@ func (c *TimedCache[T]) invalidateTask() {
 			expired := time.Now().After(c.expiration)
 			c.mutex.Unlock()
 			if expired {
-				c.Invalidate()
+				c.invalidateExternal()
 			}
 		case <-c.stopChan:
 			return
-		default:
-			// Check if TTL is still non-zero and return if true, to avoid possible memory leaks
-			if c.ttl == 0 {
-				return
-			}
 		}
 	}
 }
 
 func (c *TimedCache[T]) Stop() {
+	if !atomic.CompareAndSwapUint32(&c.stopped, cacheStateActive, cacheStateStopping) {
+		return
+	}
 	close(c.stopChan)
+	c.stopWg.Wait()
 }
 
 func (c *TimedCache[T]) Invalidate() {
+	if atomic.LoadUint32(&c.stopped) != cacheStateActive {
+		return
+	}
+	c.invalidateExternal()
+}
+
+func (c *TimedCache[T]) invalidateExternal() {
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
+	c.invalidateLocked()
+}
 
+func (c *TimedCache[T]) invalidateLocked() {
+	if atomic.LoadUint32(&c.stopped) != cacheStateActive {
+		return
+	}
 	c.isSet = false
-	close(c.stopChan)
-	c.stopChan = make(chan struct{})
 }

--- a/core/pkg/policyhandler/cache_bug_test.go
+++ b/core/pkg/policyhandler/cache_bug_test.go
@@ -1,0 +1,255 @@
+package policyhandler
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"testing"
+	"time"
+)
+
+func TestTimedCacheInvalidateTwiceNoPanic(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+	cache.Set("test-value")
+
+	cache.Invalidate()
+	cache.Invalidate()
+}
+
+func TestTimedCacheInvalidateAfterStopNoPanic(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+	cache.Set("test-value")
+
+	cache.Stop()
+	cache.Invalidate()
+}
+
+func TestTimedCacheRaceInvalidateAndStopNoPanic(t *testing.T) {
+	var panicCount atomic.Int64
+
+	for i := 0; i < 100; i++ {
+		cache := NewTimedCache[string](1 * time.Hour)
+		cache.Set("test")
+
+		var innerWg sync.WaitGroup
+		innerWg.Add(2)
+
+		var panicErr error
+		var panicMu sync.Mutex
+
+		go func() {
+			defer innerWg.Done()
+			defer func() {
+				if r := recover(); r != nil {
+					panicMu.Lock()
+					panicErr = fmt.Errorf("%v", r)
+					panicMu.Unlock()
+				}
+			}()
+			cache.Invalidate()
+		}()
+
+		go func() {
+			defer innerWg.Done()
+			defer func() {
+				if r := recover(); r != nil {
+					panicMu.Lock()
+					panicErr = fmt.Errorf("%v", r)
+					panicMu.Unlock()
+				}
+			}()
+			cache.Stop()
+		}()
+
+		innerWg.Wait()
+
+		if panicErr != nil {
+			panicCount.Add(1)
+		}
+	}
+
+	if panicCount.Load() > 0 {
+		t.Errorf("%d panics occurred during race", panicCount.Load())
+	}
+}
+
+func TestTimedCacheStopIdempotent(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+	cache.Set("test-value")
+
+	cache.Stop()
+	cache.Stop()
+}
+
+func TestTimedCacheStopWaitsForGoroutine(t *testing.T) {
+	cache := NewTimedCache[string](10 * time.Millisecond)
+	cache.Set("test-value")
+
+	goroutineExited := make(chan bool, 1)
+
+	go func() {
+		cache.Stop()
+		goroutineExited <- true
+	}()
+
+	select {
+	case <-goroutineExited:
+	case <-time.After(500 * time.Millisecond):
+		t.Error("Stop() did not return after 500ms; goroutine may not have exited")
+	}
+}
+
+func TestTimedCacheSetCallsInvalidateOnExpiry(t *testing.T) {
+	cache := NewTimedCache[string](10 * time.Millisecond)
+
+	cache.Set("test-value")
+	val, ok := cache.Get()
+	if !ok || val != "test-value" {
+		t.Errorf("Expected to get 'test-value', got '%s', ok=%v", val, ok)
+	}
+
+	deadline := time.Now().Add(500 * time.Millisecond)
+	for time.Now().Before(deadline) {
+		_, ok = cache.Get()
+		if !ok {
+			return
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+
+	t.Error("Cache did not expire after 500ms")
+}
+
+func TestTimedCacheBasicSetGet(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+
+	_, ok := cache.Get()
+	if ok {
+		t.Error("Expected cache to be empty initially")
+	}
+
+	cache.Set("test-value")
+	val, ok := cache.Get()
+	if !ok || val != "test-value" {
+		t.Errorf("Expected 'test-value', got '%s', ok=%v", val, ok)
+	}
+}
+
+func TestTimedCacheZeroTTLNoGoroutine(t *testing.T) {
+	cache := NewTimedCache[string](0)
+
+	cache.Set("test-value")
+
+	_, ok := cache.Get()
+	if ok {
+		t.Error("Expected cache to be disabled with TTL=0")
+	}
+
+	cache.Stop()
+	cache.Invalidate()
+}
+
+func TestTimedCacheMultipleInvalidateBeforeExpiry(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+
+	cache.Set("value1")
+	cache.Invalidate()
+	cache.Invalidate()
+	cache.Invalidate()
+
+	val, ok := cache.Get()
+	if ok {
+		t.Error("Expected cache to be invalid after Invalidate()")
+	}
+	_ = val
+}
+
+func TestTimedCacheCacheReusableAfterInvalidate(t *testing.T) {
+	cache := NewTimedCache[string](1 * time.Hour)
+
+	cache.Set("value1")
+	val, ok := cache.Get()
+	if !ok || val != "value1" {
+		t.Errorf("Expected 'value1', got '%s', ok=%v", val, ok)
+	}
+
+	cache.Invalidate()
+
+	_, ok = cache.Get()
+	if ok {
+		t.Error("Expected cache to be empty after Invalidate()")
+	}
+
+	cache.Set("value2")
+	val, ok = cache.Get()
+	if !ok || val != "value2" {
+		t.Errorf("Expected 'value2' after re-Set, got '%s', ok=%v", val, ok)
+	}
+}
+
+func TestTimedCacheGetReturnsStaleValue(t *testing.T) {
+	cache := NewTimedCache[string](10 * time.Millisecond)
+	cache.Set("original")
+
+	deadline := time.Now().Add(500 * time.Millisecond)
+	for time.Now().Before(deadline) {
+		_, ok := cache.Get()
+		if !ok {
+			return
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+
+	t.Error("Cache did not expire after 500ms")
+}
+
+func TestTimedCacheExpirationWorksAfterInvalidate(t *testing.T) {
+	cache := NewTimedCache[string](10 * time.Millisecond)
+
+	cache.Set("value1")
+	_, ok := cache.Get()
+	if !ok {
+		t.Fatal("Expected value1 to be set")
+	}
+
+	cache.Invalidate()
+	_, ok = cache.Get()
+	if ok {
+		t.Fatal("Expected cache to be empty after Invalidate()")
+	}
+
+	cache.Set("value2")
+
+	deadline := time.Now().Add(500 * time.Millisecond)
+	for time.Now().Before(deadline) {
+		_, ok = cache.Get()
+		if !ok {
+			return
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+
+	t.Error("TTL expiration did not fire after Invalidate()")
+}
+
+func TestTimedCacheMultipleInvalidatesDontBreakTTL(t *testing.T) {
+	cache := NewTimedCache[string](10 * time.Millisecond)
+
+	cache.Set("v1")
+	cache.Invalidate()
+	cache.Invalidate()
+	cache.Set("v2")
+	cache.Invalidate()
+	cache.Set("v3")
+
+	deadline := time.Now().Add(500 * time.Millisecond)
+	for time.Now().Before(deadline) {
+		_, ok := cache.Get()
+		if !ok {
+			return
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+
+	t.Error("TTL expiration did not fire after multiple Invalidate() calls")
+}

